[processor/deltatocumulative]: expire stale series

tombrk opened the pull request:
**Description:** Removes stale series from tracking (and thus frees their memory) using staleness logic from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31089

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30705, https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31016

**Testing:** `TestExpiry`
**Documentation:** README updated

tombrk committed:
deltatocumulative: use exp/metrics/identity

tombrk committed:
deltatocumulative: use exp/metrics/streams.Map

tombrk committed:
*: cherry-pick staleness

tombrk committed:
deltatocumulative: staleness map

uses the staleness tracking map.

for now does not actually expire entries, until lifecycle management is
in place

tombrk committed:
metrics/staleness: synchronization

adds mutex-sync to the staleness struct.
this is especially important for item expiration, which modifies the pq
and map serially. intermittent changes by other routines would lead to
data inconsistencies here.

tombrk committed:
deltatocumulative: expire stale series

sh0rez commented:
@RichieSams please take a look at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31337/commits/7745b7e7bc75a859018143b990ccfb049fd41b5a, had to add mutex-sync to your staleness code.
I think its worth already picking that into your bring-up PR

tombrk committed:
deltatocumulative: context lifecycle

converts from Start() methods to context based cancellation

RichieSams commented:
> @RichieSams please take a look at [7745b7e](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/7745b7e7bc75a859018143b990ccfb049fd41b5a), had to add mutex-sync to your staleness code. I think its worth already picking that into your bring-up PR

IMO the mutex should be outside the structure. Taking a lock for each and every operation will get expensive fast, for the amount of data entrypoints that the processors will handle. IE, the user should take a lock once, do all the operations, and then release the lock.

RichieSams commented:
I do like your decision to pull the `Map` instance into a private variable and not expose it. That should prevent users from accidentally using it directly.

sh0rez commented:
@RichieSams I agree. The reason for the extensive locking here is to synchronize with the out-of-band "pruning" routine, which needs to have exclusive ownership while it runs. I do have a new idea however how to separate those, will try it out :+1:

tombrk committed:
stale: separate tracking, expiry and locking

tombrk committed:
*: forward to main

tombrk committed:
*: post-merge fixes

tombrk committed:
*: forward to main

tombrk committed:
staleness: remove errornous time.Now

tombrk committed:
*: linter fixes

tombrk committed:
*: docs

tombrk committed:
*: linter fixes

tombrk committed:
*: changelog

sh0rez commented:
@RichieSams @djaglowski @jpkrohling please take a look :)

RichieSams commented on processor/deltatocumulativeprocessor/internal/streams/mutex.go:
```diff
@@ -0,0 +1,47 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+)
+
+type mtx[T any, M Map[T]] struct {
```
IMO, locking per operation is going to add a large amount of overhead. And doing it with this wrap structure adds a decent amount of cognitive load for folks reading the code. It would be much simpler to take a lock at a higher level. IE at `ConsumeMetrics()` or `Aggregate()`

RichieSams commented on processor/deltatocumulativeprocessor/internal/delta/delta.go:
```diff
@@ -4,42 +4,57 @@
 
-func Numbers() streams.Aggregator[data.Number] {
-	return construct[data.Number]()
+func construct[D data.Point[D]](ctx context.Context, opts Options) streams.Aggregator[D] {
```
Minor. Can this be `newAggregator()` to match the constructor style of other golang interfaces / structs?

RichieSams commented on internal/exp/metrics/staleness/staleness.go:
```diff
@@ -55,13 +61,20 @@ func (s *Staleness[T]) Items() func(yield func(identity.Stream, T) bool) bool {
+	return s.items.Len()
+}
+
+func (s *Staleness[T]) Next() (identity.Stream, time.Time) {
```
Could we name this `NextExpiry()` and just return the time?

RichieSams commented on internal/exp/metrics/staleness/map.go:
```diff
@@ -7,24 +7,11 @@ import (
 	"time"
```
This can be deleted, since it's moved to streams. Correct? 

And HashMap is identical to RawMap

RichieSams commented on processor/deltatocumulativeprocessor/internal/streams/expiry.go:
```diff
@@ -0,0 +1,68 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/streams"
+)
+
+func ExpireAfter[T any](ctx context.Context, items streams.Map[T], ttl time.Duration) streams.Map[T] {
```
IMO, wrapping the expiry functionality in a struct obfuscates how things work together. And makes it harder to do things like mutex sharing. You're wrapping some functions, but then exposing other base functions.

It would be clearer to just have a plain method in say, Processor() which did this behavior, and took in pointers

tombrk committed:
*: high-level locking

Moves all locking to the upper level, resulting a much clearer overall
picture.

Refactors delta.Accumulator to the streams.Map interface, allowing
higher composability

sh0rez commented:
@RichieSams that was really good feedback!

I moved locking centrally into the highest level processor. Stale expiry is also called from there, leading to a very clear locking flow without hidden complexity.

Also refactored the `delta.Accumulator` to be a `streams.Map` itself (with the `Store` operation doing the accumulation), making it possible to wrap the Accumulator in Staleness, and further composability in the future.

tombrk committed:
*: linter fixes

tombrk committed:
staleness: linter fixes

tombrk committed:
streams: TestExpiry

Tests expiry functionality under real conditions, using a faked clock

tombrk committed:
*: forward to main

sh0rez commented:
@RichieSams @djaglowski I think this is ready for a final review. Added docs and a thorough test-case.

jpkrohling commented:
I guess the new release caused conflicts with this one. Once the conflicts are resolved, I'll merge this.

tombrk committed:
*: forward to main

tombrk committed:
clock: sync fake

RichieSams commented on internal/exp/metrics/staleness/staleness.go:
```diff
@@ -55,13 +61,24 @@ func (s *Staleness[T]) Items() func(yield func(identity.Stream, T) bool) bool {
 	now := NowFunc()
-
 	for {
+		if s.Len() == 0 {
```
good catch

RichieSams commented on processor/deltatocumulativeprocessor/internal/streams/expiry.go:
```diff
@@ -0,0 +1,60 @@
+		case n == 0:
+			<-e.sig
+		case n > 0:
+			expires := e.Staleness.Next().Add(e.Max)
```
This has the potential of not being protected by the lock. Since you release the lock on return of this function, but this is a go-routine.

IMO, this whole struct could go away, and the Processor could keep `Staleness` structs at the top level instead of maps. While "elegant", I'm not sure the elegance of sleeping "exactly" to the next expiry outweighs the potential bugs like this race, and the extra logic you had to add to Store() to make sure you can preemptively wake things up. 

The handling of `n` and calculation of `expires` needs to happen outside the goroutine, so it's protected by the lock

RichieSams replied:
My vote would be to replace this logic with a simple Ticker() every say 5 minutes. (We *could* make that a config value if want)

sh0rez replied:
that's a really good catch and it says quite something about complexity that no one noticed it before.

The issue I see with a timer is that streams possibly live up to `2*max-1` seconds if the timer hits in a bad moment

RichieSams replied:
I believe it would be `max+tickerInterval` seconds. IMO in most real-world scenarios `max` will be >> the ticker value. IE, `max` is 1 hour, and ticker is 5 minutes. So a series lasting 1h 4.9999min isn't the end of the world.

There is precedent for this kind of expiry schema in other metrics tools, like [mtail](https://github.com/google/mtail/blob/main/docs/Deploying.md#setting-garbage-collection-intervals)

We *could* allow the tickerInterval to be configurable. And add config validation that `max` > `tickerInterval`. With a big doc / warning that says that `tickerInterval should be much smaller than max to ensure a bad lineup of the tick doesn't allow series to last too long`

RichieSams commented on processor/deltatocumulativeprocessor/internal/streams/expiry.go:
```diff
@@ -0,0 +1,60 @@
+
+	// "try-send" to notify possibly sleeping expiry routine
+	select {
+	case e.sig <- struct{}{}:
```
This means that for every store, we're potentially starting a new "sleep" go-routine for expiry. Lots of stores quickly in a row could build up many go-routines that are all sleeping. Since each loop of creates a new one.

sh0rez replied:
are you sure?

afaict there is only one routine (started in Processor.Start). the linked code here is a try-send operation that either sends an empty struct into the channel (if the other routine happens to listen) or just does nothing at all

RichieSams replied:
In processor.Start() we launch a goroutine that does this:

```
for {
	p.mtx.Lock()
	next := p.exp.ExpireOldEntries()
	p.mtx.Unlock()

	select {
	case <-next:
	case <-p.ctx.Done():
		return
	}
}
```

`ExpireOldEntries()` is overridden above. It starts a goroutine to do the sleep. That's the goroutine I'm talking about. 

By sending a struct to the channel, you cause `next` to return early. IE, it doesn't wait for the sleep to finish / the channel to be closed. Which is the behaviour you *want*, but it also abandons that goroutine. The code will immediately loop and call `ExpireOldEntries()` again, which spawns a new goroutine. Etc.

RichieSams commented:
Sorry for the delayed final review; we got a puppy this weekend that's been eating all my time 😁 

I added some small comments about the expiry and a race condition. Otherwise, LGTM
